File: packages/chain/src/pda.ts
import {
  type Address,
  getAddressEncoder,
  getProgramDerivedAddress,
  getU16Encoder,
  getU32Encoder,
  getU64Encoder,
} from '@solana/kit'
import { WASHAPP_PROGRAM_ADDRESS } from './program.ts'

// Дзеркало `programs/washapp/src/constants.rs`; кожна адреса звірена з Rust
// через `fixtures/pda.json` (`pda.test.ts`). Числові seeds — little-endian,
// як `to_le_bytes()` на боці програми.
const CONFIG_SEED = 'config'
const MINT_SEED = 'mint'
const TREASURY_SEED = 'treasury'
const POOL_SEED = 'pool'
const VAULT_SEED = 'vault'
const SENIOR_MINT_SEED = 'senior'
const JUNIOR_MINT_SEED = 'junior'
const LOSS_SEED = 'loss'
const PROTECTION_SEED = 'protection'
const PVAULT_SEED = 'pvault'
const SELLER_SEED = 'seller'
const CONTRACT_SEED = 'contract'

type Seed = string | Uint8Array

async function derive(seeds: readonly Seed[]): Promise<Address> {
  const [pda] = await getProgramDerivedAddress({
    programAddress: WASHAPP_PROGRAM_ADDRESS,
    seeds: [...seeds],
  })
  return pda
}

export function addressBytes(value: Address): Uint8Array {
  return new Uint8Array(getAddressEncoder().encode(value))
}

// Кодери kit самі відкидають значення поза межами типу (u16/u32/u64) —
// id пулу понад 65 535 не стане мовчки іншим пулом.
const u16 = (value: number): Uint8Array => new Uint8Array(getU16Encoder().encode(value))
const u32 = (value: number): Uint8Array => new Uint8Array(getU32Encoder().encode(value))
const u64 = (value: bigint): Uint8Array => new Uint8Array(getU64Encoder().encode(value))

export async function configAddress(): Promise<Address> {
  return derive([CONFIG_SEED])
}

export async function mintAddress(): Promise<Address> {
  return derive([MINT_SEED])
}

export async function treasuryAddress(): Promise<Address> {
  return derive([TREASURY_SEED])
}

export async function poolAddress(id: number): Promise<Address> {
  return derive([POOL_SEED, u16(id)])
}

export async function vaultAddress(pool: Address): Promise<Address> {
  return derive([VAULT_SEED, addressBytes(pool)])
}

export async function seniorMintAddress(pool: Address): Promise<Address> {
  return derive([SENIOR_MINT_SEED, addressBytes(pool)])
}

export async function juniorMintAddress(pool: Address): Promise<Address> {
  return derive([JUNIOR_MINT_SEED, addressBytes(pool)])
}

export async function lossEventAddress(pool: Address, index: number): Promise<Address> {
  return derive([LOSS_SEED, addressBytes(pool), u32(index)])
}

export async function protectionAddress(pool: Address): Promise<Address> {
  return derive([PROTECTION_SEED, addressBytes(pool)])
}

export async function pvaultAddress(pool: Address): Promise<Address> {
  return derive([PVAULT_SEED, addressBytes(pool)])
}

export async function sellerAddress(pool: Address, owner: Address): Promise<Address> {
  return derive([SELLER_SEED, addressBytes(pool), addressBytes(owner)])
}

// Nonce — `bigint`: лічильник `ProtectionPool.contracts` — u64, `Number` його не вміщає.
export async function contractAddress(
  pool: Address,
  buyer: Address,
  nonce: bigint,
): Promise<Address> {
  return derive([CONTRACT_SEED, addressBytes(pool), addressBytes(buyer), u64(nonce)])
}

File: programs/washapp/tests/common/pda.rs
// Деривація всіх PDA програми з одного місця: тести US1–US4 беруть адреси
// звідси, а `tests/pda.rs` пише з них `fixtures/pda.json` для `pda.ts`.
use solana_pubkey::Pubkey;
use washapp::constants::{
    CONFIG_SEED, CONTRACT_SEED, JUNIOR_MINT_SEED, LOSS_SEED, MINT_SEED, POOL_SEED, PROTECTION_SEED,
    PVAULT_SEED, SELLER_SEED, SENIOR_MINT_SEED, TREASURY_SEED, VAULT_SEED,
};

fn find(seeds: &[&[u8]]) -> (Pubkey, u8) {
    Pubkey::find_program_address(seeds, &washapp::ID)
}

pub fn config() -> (Pubkey, u8) {
    find(&[CONFIG_SEED])
}

pub fn mint() -> (Pubkey, u8) {
    find(&[MINT_SEED])
}

pub fn treasury() -> (Pubkey, u8) {
    find(&[TREASURY_SEED])
}

pub fn pool(id: u16) -> (Pubkey, u8) {
    find(&[POOL_SEED, &id.to_le_bytes()])
}

// Vault, мінти траншів і pvault ключуються адресою пулу, не його id:
// їхні seeds ніколи не потребують розбору id з байтів.
pub fn vault(pool: &Pubkey) -> (Pubkey, u8) {
    find(&[VAULT_SEED, pool.as_ref()])
}

pub fn senior_mint(pool: &Pubkey) -> (Pubkey, u8) {
    find(&[SENIOR_MINT_SEED, pool.as_ref()])
}

pub fn junior_mint(pool: &Pubkey) -> (Pubkey, u8) {
    find(&[JUNIOR_MINT_SEED, pool.as_ref()])
}

pub fn loss_event(pool: &Pubkey, index: u32) -> (Pubkey, u8) {
    find(&[LOSS_SEED, pool.as_ref(), &index.to_le_bytes()])
}

pub fn protection(pool: &Pubkey) -> (Pubkey, u8) {
    find(&[PROTECTION_SEED, pool.as_ref()])
}

pub fn pvault(pool: &Pubkey) -> (Pubkey, u8) {
    find(&[PVAULT_SEED, pool.as_ref()])
}

pub fn seller(pool: &Pubkey, owner: &Pubkey) -> (Pubkey, u8) {
    find(&[SELLER_SEED, pool.as_ref(), owner.as_ref()])
}

pub fn contract(pool: &Pubkey, buyer: &Pubkey, nonce: u64) -> (Pubkey, u8) {
    find(&[
        CONTRACT_SEED,
        pool.as_ref(),
        buyer.as_ref(),
        &nonce.to_le_bytes(),
    ])
}
